### PR TITLE
Initial TypeScript support

### PR DESCRIPTION
**Pull request by gantony**
Use typescript-compiler npm package as the engine to compile typescript code.

It currently only support one .ts input file and generate a corresponding .js file.

I still have an issue with the supported version: I'm too stringent with the required version: I hardcoded 1.4.1-2 as a requirement but the code should work with any version.
I just don't know how to name my files to avoid the exception in accord.load(). Any help on this would be appreciated.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,6 +22,8 @@
   },
   "devDependencies": {
     "LiveScript": "latest",
+    "typescript": "latest",
```
Comment (davej) [inline package.json]: Looks like `typescript-compiler` already includes `typescript` as a git submodule so I don't think this line is needed.

**Comment by jenius** [comment]
This is good! Nice work, and thank you :grinning: just a couple comments to clean things up and we'll be ready to merge and ship

**Comment by gantony** [comment]
Ok, should be good to go this time. Not too happy about forcing 1.4.1 in `package.json`, but seems to be the only way to make it work, and it's simple enough to update if needed.

Thanks for the feedback :smile:

**Comment by davej** [comment]
@gantony Did you try `"^1.4"`? It should work fine and it will be compatible with future 1.x releases.

**Comment by gantony** [comment]
@davej I did (see details in commit message). Tried again with "~1.4" after pushing but no luck. Maybe there is something wrong with my setup too...

**Comment by davej** [comment]
Hmmm, odd, works for me with both npm2 and npm 3.

accord git:(pr/130) ➜  npm install typescript-compiler@^1.4
accord@0.20.5 /Users/dave/github/accord
└── typescript-compiler@1.4.1

**Comment by gantony** [comment]
Ok, seems like I'm using an ancient version of npm :)

    antony@antony-ThinkPad-T410:~/dev/accord$ npm install typescript-compiler@^1.4
    npm http GET https://registry.npmjs.org/typescript-compiler
    npm http 304 https://registry.npmjs.org/typescript-compiler
    typescript-compiler@1.4.1-2 node_modules/typescript-compiler
    antony@antony-ThinkPad-T410:~/dev/accord$ npm -v
    1.3.10
    antony@antony-ThinkPad-T410:~/dev/accord$

**Comment by davej** [comment]
Aha, that explains it. Accord supports Node 0.10 (and npm 1.x) so perhaps keep it hardcoded for the moment. When [this](https://github.com/theblacksmith/typescript-compiler/issues/15) is fixed, then we can use a more flexible version range.

**Comment by gantony** [comment]
I guess it's an improvement from the previous commit with `1.4.1-2` hardcoded in the adapter too :grinning:

**Comment by jenius** [comment]
Ok so let's use `^1.4` for now, then once the linked issue is resolved we can change it to be more flexible :grinning:

**Comment by gantony** [comment]
@jenius I assume you mean let's use `1.4.1` for now as it will work for npm version 1, 2 and 3, and only switch to `^1.4` once the linked issue is fixed (since `^1.4` does not work with npm version 1, but only 2 and 3). If so, the branch can be merged as it is now.

**Comment by jenius** [comment]
@gantony tbh I am entirely ok with not supporting npm v1, and moving accord up in version support from 0.10. This is a very old version of npm and node and it's about time to upgrade!

**Comment by gantony** [comment]
Typescript tests are passing on my machine after upgrading to npm 3. They are failing on CI since it's using npm version 1.4.29

**Comment by davej** [comment]
@gantony Yeah, `.travis.yml` will have to be bumped to `0.12`, I assume @jenius is ok with this if support is being dropped from `0.10`. Probably good to add the latest LTE (`4.x`) also.

```yaml
language: node_js
node_js:
  - '0.12'
  - '4'
after_script:
  - npm run coveralls
```

**Comment by jenius** [comment]
Yep, I'm ok with this. While we're at it, we should make sure it's working with v5 as well

**Comment by jenius** [comment]
Ok, rebase on top of master and let's see how this looks. Latest version only tests v4 and up. Time to upgrade guys :grinning:

**Comment by gantony** [comment]
All green, seems like we can close this one :)

**Comment by jenius** [comment]
Something is wrong here with the commit. I think you did some type of merge rather than rebasing on top of master. All the typescript code that actually did anything was removed. Also the commit message is not exactly clear. I might roll back and try again? You are looking to fetch and rebase on top of the upstream here :grinning:

**Comment by gantony** [comment]
Yeah I fetched and merged instead of rebasing. The git graph looks funny but the changes are there, since we can see that the typescripts tests are passing.

Anyway, I'll revert the merge and rebase instead if it's better for merging the pull request.

**Comment by gantony** [comment]
Ok, rewrote a bit of history to replace the merge with a rebase. Hope it's OK this time.

**Comment by jenius** [comment]
Looks great! Thanks for your hard work on this :+1: 

So a couple more things:
1) Could you add a file to the `docs` folder for typescript, with a short description of any options, etc?
2) Could you add typescript to the list of supported languages in the readme?

After this we should be ready to merge & ship :grin:

**Comment by gantony** [comment]
Ok, this time I think we should be good to go.

**Comment by jenius** [comment]
Whoo! :tada: - lookgs good to me. Let's make this happen. Thanks so much for really working super hard at getting this to happen @gantony. I'll release now so it can be used immediately :grin:

**Comment by matthew-dean** [comment]
Ohhhh shiiiiiiiii

I LOVE ALL OF YOU

**Comment by matthew-dean** [comment]
Hang on... the docs say this wraps the typescript-compiler command-line wrapper.

Doesn't that imply that a user has to have the typescript command-line tools installed in order for this to work? Why not call the TypeScript Compiler API directly? Am I missing something? https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#transpiling-single-file

**Comment by matthew-dean** [comment]
Wouldn't this also only directly write a file to disk rather than returning strings for file content and source map?

**Comment by jenius** [comment]
@matthew-dean it works for everything. tests are in place and passing. give it a shot!

**Pull request merged**